rkt/api_service.go

```diff
@@ -691,8 +716,86 @@ func (s *v1AlphaAPIServer) InspectImage(ctx context.Context, request *v1alpha.In
 	return &v1alpha.InspectImageResponse{Image: image}, nil
 }
 
+// LogsStreamWriter is a wrapper around a GRPC streaming server.
+// Implements io.Writer interface.
+type LogsStreamWriter struct {
+	server v1alpha.PublicAPI_GetLogsServer
+}
+
+func (sw LogsStreamWriter) Write(b []byte) (int, error) {
+	if err := sw.server.SendMsg(b); err != nil {
+		return 0, err
+	}
+	return len(b), nil
+}
+
 func (s *v1AlphaAPIServer) GetLogs(request *v1alpha.GetLogsRequest, server v1alpha.PublicAPI_GetLogsServer) error {
-	return fmt.Errorf("not implemented yet")
+	uuid, err := types.NewUUID(request.PodId)
+	if err != nil {
+		return err
+	}
+	pod, err := getPod(uuid)
+	if err != nil {
+		return err
+	}
+	stage1Path := "stage1/rootfs"
+	if pod.usesOverlay() {
+		stage1TreeStoreID, err := pod.getStage1TreeStoreID()
+		if err != nil {
+			return err
+		}
+		stage1Path = fmt.Sprintf("/overlay/%s/upper/", stage1TreeStoreID)
+	}
+	path := filepath.Join(getDataDir(), "/pods/run/", request.PodId, stage1Path, "/var/log/journal/")
+	jconf := sdjournal.JournalReaderConfig{
+		Path: path,
+	}
+	if request.AppName != "" {
+		acname, err := types.NewACName(request.AppName)
+		if err != nil {
+			return err
+		}
+		jconf.Matches = []sdjournal.Match{
+			{
+				Field: sdjournal.SD_JOURNAL_FIELD_SYSLOG_IDENTIFIER,
+				Value: stage1common.ServiceUnitName(*acname),
+			},
+		}
+	}
+	if request.SinceTime != 0 {
+		t := time.Unix(request.SinceTime, 0)
+		jconf.Since = time.Since(t)
+	}
+	if request.Lines != 0 {
+		jconf.NumFromTail = uint64(request.Lines)
+	}
+
+	s.mu.Lock()
+	jr, ok := s.journalReaders[request.PodId]
+	if !ok {
+		journalReader, err := sdjournal.NewJournalReader(jconf)
+		if err != nil {
+			s.mu.Unlock()
+			return err
+		}
+		jr = &mtJournalReader{JournalReader: journalReader}
+		s.journalReaders[request.PodId] = jr
+	}
+	jr.mu.Lock()
+	defer jr.mu.Unlock()
```

Comment:
I agree that the API service should support several GetLogs() calls on the same pod at the same time, specially because it has this "Follow" feature, so the method call could run for a very long time.